### PR TITLE
Don't apply the plugin's reader wrapper in can_match phase

### LINKED ISSUE #46817: DLS search performance/canMatch impact
During `IndexShard.acquireSearcher`, the `readerWrapper` is applied, which populates the DLS bitsets. This causes a performance issue, since `IndexShard.acquireSearcher` is called during the `canMatch` phase. In this context it seems unnecessary, since we are only after whether the shard could match or not, which does not require DLS to kick in. When searching for a short time range in a large number of time based indices, this causes a performance impact which seems unnecessary.

Concrete observations: A query that takes 30ms can take 26s for a DLS filtered user. The user has many DLS roles applied, which seems to increase the impact of the issue. The hot threads are dominated by variants of this stack trace:

```
app//org.apache.lucene.search.DisjunctionDISIApproximation.nextDoc(DisjunctionDISIApproximation.java:55)
app//org.apache.lucene.search.DisjunctionDISIApproximation.nextDoc(DisjunctionDISIApproximation.java:55)
app//org.apache.lucene.util.BitSet.or(BitSet.java:95)
app//org.apache.lucene.util.FixedBitSet.or(FixedBitSet.java:271)
app//org.apache.lucene.util.BitSet.of(BitSet.java:41)
org.elasticsearch.xpack.core.security.authz.accesscontrol.DocumentSubsetBitsetCache.lambda$getBitSet$2(DocumentSubsetBitsetCache.java:153)
org.elasticsearch.xpack.core.security.authz.accesscontrol.DocumentSubsetBitsetCache$$Lambda$5134/0x000000080204e840.load(Unknown Source)
app//org.elasticsearch.common.cache.Cache.computeIfAbsent(Cache.java:433)
org.elasticsearch.xpack.core.security.authz.accesscontrol.DocumentSubsetBitsetCache.getBitSet(DocumentSubsetBitsetCache.java:135)
org.elasticsearch.xpack.core.security.authz.accesscontrol.DocumentSubsetReader.<init>(DocumentSubsetReader.java:160)
org.elasticsearch.xpack.core.security.authz.accesscontrol.DocumentSubsetReader.<init>(DocumentSubsetReader.java:34)
org.elasticsearch.xpack.core.security.authz.accesscontrol.DocumentSubsetReader$DocumentSubsetDirectoryReader$1.wrap(DocumentSubsetReader.java:120)
app//org.apache.lucene.index.FilterDirectoryReader$SubReaderWrapper.wrap(FilterDirectoryReader.java:62)
app//org.apache.lucene.index.FilterDirectoryReader.<init>(FilterDirectoryReader.java:91)
org.elasticsearch.xpack.core.security.authz.accesscontrol.DocumentSubsetReader$DocumentSubsetDirectoryReader.<init>(DocumentSubsetReader.java:116)
org.elasticsearch.xpack.core.security.authz.accesscontrol.DocumentSubsetReader.wrap(DocumentSubsetReader.java:38)
org.elasticsearch.xpack.core.security.authz.accesscontrol.SecurityIndexReaderWrapper.apply(SecurityIndexReaderWrapper.java:86)
org.elasticsearch.xpack.core.security.authz.accesscontrol.SecurityIndexReaderWrapper.apply(SecurityIndexReaderWrapper.java:42)
app//org.elasticsearch.index.shard.IndexShard.wrapSearcher(IndexShard.java:1262)
app//org.elasticsearch.index.shard.IndexShard.acquireSearcher(IndexShard.java:1240)
app//org.elasticsearch.index.shard.IndexShard.acquireSearcher(IndexShard.java:1224)
app//org.elasticsearch.search.SearchService.createSearchContext(SearchService.java:632)
app//org.elasticsearch.search.SearchService.canMatch(SearchService.java:1007)
app//org.elasticsearch.search.SearchService.canMatch(SearchService.java:1020)
app//org.elasticsearch.action.search.SearchTransportService.lambda$registerRequestHandler$14(SearchTransportService.java:396)
app//org.elasticsearch.action.search.SearchTransportService$$Lambda$3005/0x0000000801b4d840.messageReceived(Unknown Source)
```

These all run directly in the transport thread.

Encountered on ES v7.3.0, running on linux OS.

### PR DESCRIPTION
**Pull request by jimczi**
This change modifies the local execution of the `can_match` phase to **not** apply
the plugin's reader wrapper (if it is configured) when acquiring the searcher.
We must ensure that the phase runs quickly and since we don't know the cost
of applying the wrapper it is preferable to avoid it entirely. The can_match
phase can aford false positives so it is also safe for the builtin plugins
that use this functionality.

Closes #46817

### CODE REVIEW

**Review by henningandersen — COMMENTED** [review]
Thanks @jimczi , this seems like an easy and good approach. I left a single comment that I think need to be addressed.

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 1]
```diff
@@ -1012,8 +1012,12 @@ public AliasFilter buildAliasFilter(ClusterState state, String index, Set<String
      */
     public boolean canMatch(ShardSearchRequest request) throws IOException {
         assert request.searchType() == SearchType.QUERY_THEN_FETCH : "unexpected search type: " + request.searchType();
-        try (DefaultSearchContext context = createSearchContext(request, defaultSearchTimeout, false, "can_match")) {
```
[the remark this answers is not in the document]
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/SearchService.java]: Indeed, sorry for wasting your time ;). I pushed a [change](https://github.com/elastic/elasticsearch/pull/47816/commits/1f5941a881fce0b0de066d35cd1c127c4af868bc) to restore the request rewrite.

File: server/src/test/java/org/elasticsearch/search/SearchServiceTests.java [hunk 2]
```diff
@@ -582,6 +623,13 @@ public void testCanMatch() throws IOException {
         searchRequest.source(new SearchSourceBuilder().query(new MatchNoneQueryBuilder()));
         assertFalse(service.canMatch(new ShardSearchRequest(OriginalIndices.NONE, searchRequest, indexShard.shardId(), 1,
             new AliasFilter(null, Strings.EMPTY_ARRAY), 1f, -1, null, null)));
+        assertEquals(numWrapReader, numWrapInvocations.get());
+
+        // make sure that the wrapper is called when we create the context is actually created
```
Comment (jpountz) [inline server/src/test/java/org/elasticsearch/search/SearchServiceTests.java]: this sounds like it's mixing up two sentences?

**Review by jpountz — APPROVED** [review]
I think this is a good short-term fix.

It's not the first time (neither the second time...) that we get bitten by the fact that BaseCompositeReaderWrapper computes `numDocs` eagerly. I think we should look into fixing this and hopefully remove this workaround.

**Review by henningandersen — APPROVED** [review]
LGTM.

**Comment by jimczi** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by jimczi** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by jimczi** [comment]
@elasticmachine update branch

**Pull request merged**